File: src/ui/ModulesWriter.java
package ui;

import com.google.gson.Gson;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

public class ModulesWriter {

    public ModulesWriter(PackagesGeneratorUI packagesGeneratorUI) {
    }

    public void generate(String configStr) {

        Gson gson = new Gson();
        ConfigPOJO configPOJO = gson.fromJson(configStr, ConfigPOJO.class);

        writeRootFolder(configPOJO);

        for (int i = 0; i < Integer.parseInt(configPOJO.getNumModules()); i++) {
            writeModule(i, configPOJO);
        }
    }

    private void writeModule(int index, ConfigPOJO configPOJO) {
        String moduleRoot = configPOJO.getRoot() + "/module" + index +"/";
        File moduleRootFile = new File(moduleRoot);
        moduleRootFile.mkdir();

        writeLibsFolder(moduleRootFile, configPOJO);
        writeBuildGradle(moduleRootFile,configPOJO);

        PackagesWriter packagesWriter = new PackagesWriter();
        packagesWriter.writePackages(configPOJO,
                moduleRoot + "/src/main/java/");

    }

    private void writeBuildGradle(File moduleRootFile, ConfigPOJO configPOJO) {
        String libRoot = moduleRootFile + "/build.gradle/";
        BufferedWriter writer = null;
        try {
            File buildGradle = new File(libRoot);
            buildGradle.createNewFile();

            writer = new BufferedWriter(new FileWriter(buildGradle));
            writer.write(BuildGradle.TEXT);

        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                writer.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private void writeLibsFolder(File moduleRootFile, ConfigPOJO configPOJO) {
        // write libs
        String libRoot = moduleRootFile + "/libs/";
        new File(libRoot).mkdir();
    }

    private void writeRootFolder(ConfigPOJO configPOJO) {
        File root = new File(configPOJO.getRoot());

        if (!root.exists()) {
            root.mkdir();
        } else {
            root.delete();
            root.mkdir();
        }
    }
}

File: src/main/kotlin/ui/ModulesWriter.kt
/*
 *  Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package ui

import com.google.gson.Gson
import kotlinx.coroutines.experimental.Job
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch
import kotlinx.coroutines.experimental.runBlocking
import ui.generators.BuildGradleGenerator
import ui.generators.PackagesGenerator
import ui.generators.packages.JavaGenerator
import ui.generators.packages.KotlinGenerator
import ui.generators.project.GradleSettingsGenerator
import ui.generators.project.GradlewGenerator
import ui.generators.project.ProjectBuildGradleGenerator
import ui.models.ConfigPOJO
import ui.models.ModuleBlueprint
import utils.joinPath
import java.io.File

class ModulesWriter(private val dependencyValidator: DependencyValidator,
                    private val blueprintFactory: ModuleBlueprintFactory,
                    private val buildGradleGenerator: BuildGradleGenerator,
                    private val gradleSettingsGenerator: GradleSettingsGenerator,
                    private val projectBuildGradleGenerator: ProjectBuildGradleGenerator,
                    private val fileWriter: FileWriter) {

    fun generate(configStr: String) = runBlocking {

        val gson = Gson()
        val configPOJO = gson.fromJson(configStr, ConfigPOJO::class.java)

        if (!dependencyValidator.isValid(configPOJO)) {
            throw IllegalStateException("Incorrect dependencies")
        }

        val projectRoot = configPOJO.root.joinPath(configPOJO.projectName)

        writeRootFolder(configPOJO.root)
        writeRootFolder(projectRoot)
        GradlewGenerator.generateGradleW(projectRoot)

        val moduleBlueprints = (0 until configPOJO.numModules).map { i ->
            blueprintFactory.create(i, configPOJO, projectRoot)
        }

        gradleSettingsGenerator.generate(configPOJO.projectName, moduleBlueprints, projectRoot)
        projectBuildGradleGenerator.generate(projectRoot)

        val allJobs = mutableListOf<Job>()
        moduleBlueprints.forEach{ blueprint ->
            val job = launch {
                writeModule(blueprint, configPOJO)
            }
            allJobs.add(job)
        }
        for ((index, job) in allJobs.withIndex()) {
            println("Done writing module " + index)
            job.join()
        }
    }

    private fun writeModule(moduleBlueprint: ModuleBlueprint, configPOJO: ConfigPOJO) {
        val moduleRootPath = moduleBlueprint.root
        val moduleRoot = moduleRootPath.joinPath(moduleBlueprint.name)
        val moduleRootFile = File(moduleRoot)
        moduleRootFile.mkdir()

        writeLibsFolder(moduleRootFile)
        writeBuildGradle(moduleRootFile, moduleBlueprint)

        val packagesWriter = PackagesGenerator(JavaGenerator(fileWriter), KotlinGenerator(fileWriter))

        // TODO stopped here add index
        packagesWriter.writePackages(configPOJO, moduleBlueprint.index,
                moduleRoot + "/src/main/java/", File(moduleRootPath))
    }

    private fun writeBuildGradle(moduleRootFile: File, moduleBlueprint: ModuleBlueprint) {
        val libRoot = moduleRootFile.toString() + "/build.gradle/"
        val content = buildGradleGenerator.create(moduleBlueprint)
        fileWriter.writeToFile(content, libRoot)
    }

    private fun writeLibsFolder(moduleRootFile: File) {
        // write libs
        val libRoot = moduleRootFile.toString() + "/libs/"
        File(libRoot).mkdir()
    }

    private fun writeRootFolder(projectRoot: String) {
        val root = File(projectRoot)

        if (!root.exists()) {
            root.mkdir()
        } else {
            root.delete()
            root.mkdir()
        }
    }
}
